witchcraft/service_discovery.go

```diff
@@ -138,6 +182,16 @@ func (s *serviceDiscovery) serviceConfig(serviceName string) httpclient.Refresha
 	}))
 }
 
+func filterHTTPParams(in []httpclient.ClientParam) []httpclient.HTTPClientParam {
+	var out []httpclient.HTTPClientParam
+	for _, param := range in {
+		if h, ok := param.(httpclient.HTTPClientParam); ok {
```

Comment:
Clients will have to convert from the HTTPClientParam interface to the ClientParam interface in order to use this feature. This will require adding a mock implementation of ClientParam, which seems quite janky.

Could we instead accept a list of ClientOrHTTPClientParam instead? This way the client chooses how to apply to a vanilla client and how to apply to an HTTP client.

Reply:
Why does anything need to be converted? We know the exhaustive set of params because the [interface](https://github.com/palantir/conjure-go-runtime/blob/db418fba36ddde2ed5d5201e5cbe69eafa7ecb1a/conjure-go-client/httpclient/client_params.go#L33-L46) methods are private. The only one (today) that does not implement ClientParam or ClientOrHTTPClientParam is WithConfigForHTTPClient, which doesn't make sense in this case because there are already plenty of better ways to provide config. I've added a test and comment which might help.